src/sentry/replays/consumers/recording/process_recording.py

```diff
@@ -0,0 +1,191 @@
+
+import msgpack
+import sentry_sdk
+from arroyo import Partition
+from arroyo.backends.kafka.consumer import KafkaPayload
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Message, Position
+from django.conf import settings
+
+from sentry.attachments import MissingAttachmentChunks, attachment_cache
+from sentry.models import File
+from sentry.replays.consumers.recording.types import RecordingChunkMessage, RecordingMessage
+from sentry.replays.models import ReplayRecordingSegment
+from sentry.utils import json
+
+logger = logging.getLogger("sentry.replays")
+
+CACHE_TIMEOUT = 3600
+
+
+class ReplayRecordingMessageFuture(NamedTuple):
+    """
+    Map a submitted message to a Future returned by the Producer.
+    This is useful for being able to commit the latest offset back
+    to the original consumer.
+    """
+
+    message: Message[KafkaPayload]
+    future: Future[bool | None]
+
+
+class ProcessRecordingStrategy(ProcessingStrategy[KafkaPayload]):
+    def __init__(
+        self,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__closed = False
+        self.__futures: Deque[ReplayRecordingMessageFuture] = deque()
+        self.__threadpool = concurrent.futures.ThreadPoolExecutor()
+        self.__commit = commit
+
+    def poll(self) -> None:
+        self._commit_and_prune_futures()
+
+    def _process_chunk(
+        self, message_dict: RecordingChunkMessage, message: Message[KafkaPayload]
+    ) -> None:
+
+        # can't make this a future as we need to guarentee this happens before
+        # the final kafka message
+        # TODO: determine if this is acceptable at scale
+        id = message_dict["id"]
+        project_id = message_dict["project_id"]
+        chunk_index = message_dict["chunk_index"]
+        cache_key = replay_cache_id(id, project_id)
+
+        attachment_cache.set_chunk(
+            key=cache_key,
+            id=id,
+            chunk_index=chunk_index,
+            chunk_data=message_dict["payload"],
+            timeout=CACHE_TIMEOUT,
+        )
+
+    def _store(self, message_dict: RecordingMessage, recording: bytes) -> None:
+        # create a File for our recording segment.
+        recording_file_name = (
+            f"rr:{message_dict['replay_id']}:{message_dict['recording_headers']['sequence_id']}"
+        )
+
+        file = File.objects.create(
+            name=recording_file_name,
+            type="replay.recording",
+        )
+        file.putfile(
+            BytesIO(recording),
+            blob_size=settings.SENTRY_ATTACHMENT_BLOB_SIZE,
+        )
+        # associate this file with an indexable replay_id via ReplayRecordingSegment
+        ReplayRecordingSegment.objects.create(
+            replay_id=message_dict["replay_id"],
+            project_id=message_dict["project_id"],
+            sequence_id=message_dict["recording_headers"]["sequence_id"],
+            file_id=file.id,
+        )
```

Comment:
If this block fails due to GCS problems will processing become blocked? Or will the impacted replay have a 'hole' from the missing chunks?